File: user-service/apps/user-api/src/main/java/com/yellowsunn/userservice/filter/LoginUserArgumentResolver.java
package com.yellowsunn.userservice.filter;

import static com.yellowsunn.common.constant.CommonHeaderConst.USER_ID;

import com.yellowsunn.common.annotation.LoginUser;
import com.yellowsunn.common.exception.LoginRequireException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.MethodParameter;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

@Slf4j
public class LoginUserArgumentResolver implements HandlerMethodArgumentResolver {

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        return parameter.hasParameterAnnotation(LoginUser.class);
    }

    @Override
    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer,
            NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {
        String userId = webRequest.getHeader(USER_ID);
        LoginUser loginUser = parameter.getParameterAnnotation(LoginUser.class);
        if (loginUser != null && loginUser.required() && userId == null) {
            throw new LoginRequireException();
        }
        return userId;
    }
}

File: notification-service/apps/notification-api/src/main/kotlin/com/yellowsunn/notificationservice/filter/LoginUserArgumentResolver.kt
package com.yellowsunn.notificationservice.filter

import com.yellowsunn.common.annotation.LoginUser
import com.yellowsunn.common.constant.CommonHeaderConst.USER_ID
import com.yellowsunn.common.exception.LoginRequireException
import org.springframework.core.MethodParameter
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.method.support.ModelAndViewContainer

class LoginUserArgumentResolver : HandlerMethodArgumentResolver {
    override fun supportsParameter(parameter: MethodParameter): Boolean {
        return parameter.hasParameterAnnotation(LoginUser::class.java)
    }

    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?,
    ): Any? {
        val userId: Long? = webRequest.getHeader(USER_ID)?.toLongOrNull()
        val loginUser: LoginUser? = parameter.getParameterAnnotation(LoginUser::class.java)
        if (loginUser?.required == true && userId == null) {
            throw LoginRequireException()
        }
        return userId
    }
}
